Report issues with check-runs instead of only comments, and remove comments that are resolved

mattisdada opened the pull request:
TS port of the original PR #33

- Be able to switch the mode of reporting from comments (current) and to check runs
    - What is [check runs](https://github.blog/2018-12-14-introducing-check-runs-and-annotations/)? It is the built in mechanism in GitHub pull requests to add annotations from scanning tools (ie, linters). In my opinion it works a bit better than comments as it doesn't produce notifications, they are softer and show as a status on the PR
- Remove resolved comments from the PR
    - Unfortunately, the REST API cannot be used to **resolve** comments, that can only be done via the GraphQL mutations API but it doesn't have cross compatible IDs
    - Added in a prefix to each comment so they can be safely deleted - not the most reliable method
- Fixes bug that made it so the diff scanning never worked (more detail in comment)

Structurally, all of the functionality for generating comments and check-runs have been moved to their own files.

Happy to break up this PR if needed

## Check run preview
![image](https://user-images.githubusercontent.com/314500/226084975-0d2549a9-d2bb-420c-be86-8a5df242daee.png)

To test, you can use the following action:

```yaml
    - name: SFDX Scanner
      uses: mattisdada/sfdx-scan-pull-request@ts-port
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        report-mode: check-runs # or swap to `comments` to toggle between them
```

## Examples

Comments reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/3

Check runs reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/2

0.1.8 release: (this PR was trying to show the existing bug I mentioned in this PR, but there seems to be multiple issues in 0.1.8, so it's only reported one issue, the action failed)
- https://github.com/mattisdada/simple-lambda-router-ts/pull/4

Video showing the comments being deleted after resolved: (hard to show in a PR because the comment is deleted)
- https://www.youtube.com/watch?v=ds5BEyiexLc

mattisdada commented on src/index.ts:
```diff
@@ -66,14 +44,27 @@ function initialSetup() {
     severityThreshold: parseInt(getInput("severity-threshold")) || 4,
     strictlyEnforcedRules: getInput("strictly-enforced-rules"),
+    deleteResolvedComments: getInput("delete-resolved-comments") === "true",
+    target: getInput("target"),
```
This fixes the bug that made it so the diff never got scanned. Later on in this file `scannerFlags.target` always gets updated to be a list of files, either provided by the user or the action. When `filterFindingsToDiffScope` checks to see if the user has input a target, it will always be true.

By separating the input given by the user and the target used by the scanner, we can fix the issue

mitchspano replied:
What do you mean "the bug that forced every line to be reported"? Is this an existing issue?

matthew-gladman-oua replied:
Yup, since TS upgrade every line in a file is being pinged instead of just changed files

mitchspano replied:
That is interesting.... That is not something that I have experienced. Can you raise an issue with an example of this during the action's run? I would like to take a look and see if I can understand the issue and what is causing it.

mattisdada replied:
I was trying to replicate it again on the official release, but it's even more broken for me now and cannot save a second comment anymore on 0.1.8 :(

- Action: https://github.com/mattisdada/simple-lambda-router-ts/actions/runs/4487397056/jobs/7890816581
- PR: https://github.com/mattisdada/simple-lambda-router-ts/pull/4

But you can see it trying to create the erroneous comments in the failures, line 6 shouldn't have a comment applied as it wasn't changed

mitchspano replied:
So the true issue here is the order of execution where `updateScannerTarget` is setting the `scannerFlags.target` before we get to `filterFindingsToDiffScope` - at which point the `.target` is already set, so this if check will _never_ evaluate to true, thus we will never continue and _always_ report the finding - correct?

```ts
if (!isInChangedLines(violation, relevantLines) && !scannerFlags.target) {
  continue;
}
```

I wonder if we can change the `updateScannerTarget` function to return a separate `List<String>`, say `filesToScan` without mutating the `scannerFlags.target`.

mitchspano replied:
This might also explain some of the odd behavior @jbatt08 was observing in #39

matthew-gladman-oua replied:
You've hit the nail on the head. I'm pretty happy with either approach to be honest.

mitchspano replied:
I think that changing the function name from `updateScannerTarget` to `getFilesToScan` and having it return a comma separated list of files without mutation to input flags would be preferable to replicating the `.target` input to both `scannerFlags` and `input`.

We also might want to use `.join(",")` instead of trimming the trailing comma like we are doing today.

mattisdada replied:
Looks like you've already made these changes to main at some point, I'll remove the changes

matthew-gladman-oua commented on src/reporter/comments-reporter.ts:
```diff
@@ -0,0 +1,194 @@
+    }
+  }
+
+  async writeComments(
```
This function needs to be removed, it was only temporary

mitchspano commented on README.md:
```diff
@@ -42,6 +42,16 @@ Optionally provide this to scan a whole directory instead of just the diff. If t
+## `report-mode`
+
+Details which way to report issues back to GitHub, can be either:
+- `check-runs` - Shows findings as annotations on the PR
```
Is one of these the default?

mitchspano commented on src/reporter/comments-reporter.ts:
```diff
@@ -0,0 +1,194 @@
+import { BaseReporter } from "./reporter.types";
+import { ScannerViolation } from "../sfdxCli";
+
+const HIDDEN_COMMENT_PREFIX = "<!--sfdx-scanner-->";
```
Is this necessary? Would it be possible to use the existing `| Engine | Category | Rule | Severity | Type | Message | File |` header as the canonical identifier of a generated comment?

matthew-gladman-oua replied:
You're complete right, we could. The rational for a separate identifier is if the format changes, this will still work indefinitely, so we aren't tied to the formatting which may change over time

mitchspano replied:
gotcha - makes sense: good addition.

mitchspano commented on src/common.ts:
```diff
@@ -0,0 +1,96 @@
+import { ScannerViolation, ScannerViolationType } from "./sfdxCli";
```
What's the purpose of segregating this logic into `common`? We already introduced an abstract class with `reporter.types` which is where I would imagine all shared type and function definitions to be defined.

Do you have a class diagram that can show these how these classes interact and relate to each other? I am having a difficult time understanding the separation of concerns in this proposed change

mattisdada replied:
This file was slightly a remnant before the TS port, I've cleaned this up, thanks for the feedback.

I've moved everything that is only used by the reporters to `reporter.types.ts` and left everything else as is.

Common is just for any functionality that might be used across the different parts of the plugin, they could be stored in `index.ts` but I'd rather avoid circular dependencies

Dependency tree:
```mermaid
flowchart TD
    A[common.ts]
    B[annoations-reporter.ts] --> A
    C[comments-reporter.ts] --> A
    D[index.ts] --> A
    D[index.ts] --> B
    D[index.ts] --> C
```

mitchspano replied:
This graph is _sweet_ - I was unaware of mermaid until now. Thanks for sharing!

mitchspano commented:
This pull request has a few changes that appear to be in an in-progress state - like [this one](https://github.com/mitchspano/sfdx-scan-pull-request/pull/40/files#r1144024997) for example. I would also like to see some testing results of this action on a sample pull request - the previous example looks like it was left over from the `.js` version of this action. I think it might be best to take a step back, clean this up, and test it out a bit before raising a pull request.

I like the idea of providing an alternative reporting mechanism, but I am having a difficult time understanding the proposed segregation of responsibilities across these different files though.  I would appreciate it if we can create a class diagram or some more thorough documentation to help understand the proposed separation of concerns that are included in this change.

mattisdada commented:
> This pull request has a few changes that appear to be in an in-progress state - like [this one](https://github.com/mitchspano/sfdx-scan-pull-request/pull/40/files#r1144024997) for example. I would also like to see some testing results of this action on a sample pull request - the previous example looks like it was left over from the `.js` version of this action. I think it might be best to take a step back, clean this up, and test it out a bit before raising a pull request.
> 
> I like the idea of providing an alternative reporting mechanism, but I am having a difficult time understanding the proposed segregation of responsibilities across these different files though. I would appreciate it if we can create a class diagram or some more thorough documentation to help understand the proposed separation of concerns that are included in this change.

No worries, it has been well tested, just the repo I've been using for testing isn't public, but I've changed that now (it's an old project). I created PRs for each piece of functionality so you can check it out.

Comments reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/3

Check runs reporter:
- https://github.com/mattisdada/simple-lambda-router-ts/pull/2

0.1.8 release: (this PR was trying to show the existing bug I mentioned in this PR, but there seems to be multiple issues in 0.1.8, so it's only reported one issue, the action failed)
- https://github.com/mattisdada/simple-lambda-router-ts/pull/4

Video showing the comments being deleted after resolved: (hard to show in a PR because the comment is deleted)
- https://www.youtube.com/watch?v=ds5BEyiexLc

----

The rationale of the new files:

- `common.ts` - a location for commonly used functions and types that can be shared throughout the project
- `reporter/reporter.types.ts` - common types, interfaces and base class for all reporters. By having a common interface, we can minimize the number of conditionals in the code and just focus on the actions reporters need to perform, (currently just `write` and `translate`). This is similar to the factory pattern, but a little bit simplified.
- `reporter/annoations-reporter.ts` - Contains all the logic required to report to GitHub using Annotations and Check Runs
- `reporter/comments-reporter.ts` - Contains all the logic required to report to GitHub using Comments
- `index.ts` - No longer needs to concern itself with comments (or annotations/check runs), keeping the file smaller and more focused on orchestration (it's primary job), delegating reporting to GH through one of the other reporters

I hope that helps explain why I've done it this way. Happy to change directions if that's preferred though. 

**Relationship diagram of the files:**

```mermaid
flowchart TD
    A[common.ts]
    B[annoations-reporter.ts] --> A
    C[comments-reporter.ts] --> A
    D[index.ts] --> A
    D[index.ts] --> B
    D[index.ts] --> C
```

mitchspano commented on src/reporter/annoations-reporter.ts:
```diff
@@ -0,0 +1,96 @@
+  async write() {
+    console.log("Creating Check Runs using GitHub REST API...");
+
+    // TODO: Change
```
What do we need to change here?

mitchspano commented on src/reporter/reporter.types.ts:
```diff
@@ -0,0 +1,103 @@
+import { ScannerViolation } from "../sfdxCli";
+import { Context } from "@actions/github/lib/context";
+
+export type GithubCheckRun = {
```
Is this used in the `comments-reporter`? If not, it seems like it belongs in `annotations-reporter` - right?

mattisdada replied:
Understandable, but I think the representation of the comment and the annotation/check run, doesn't have anything to do with the individual reporters. Makes more sense to me for them to be imported from a types file instead of the class file 🤔 But I can shift them if you prefer

mitchspano commented on src/reporter/reporter.types.ts:
```diff
@@ -0,0 +1,103 @@
+  title: string;
+};
+
+export type GithubComment = {
```
Is this used in the `annotations-reporter`? If not, it seems like it belongs in `comments-reporter` - right?

mitchspano commented on src/reporter/comments-reporter.ts:
```diff
@@ -0,0 +1,170 @@
+const HIDDEN_COMMENT_PREFIX = "<!--sfdx-scanner-->";
+
+export class CommentsReporter extends BaseReporter<GithubComment> {
+  private performGithubRequest<T>(
```
It seems like both the `annotations` and the `comments` reporters need to generate an instance of octokit and perform a  GitHub request. Is it possible for us to move any of logic or method signature declaration up to the parent abstract class?

mattisdada replied:
Hmmm, I think trying to abstract it away is not going to simplify much. Octokit is designed in this way for a reason, a lot of the endpoints are pretty unique. If you feel strongly I can put something together, but I'm not sure what we could do that would be an actual improvement

mitchspano commented on src/reporter/reporter.types.ts:
```diff
@@ -0,0 +1,103 @@
+type GithubCommentSide = "RIGHT";
+
+export abstract class BaseReporter<T> implements Reporter {
+  protected hasHaltingError;
```
I think the act of detecting a halting error and calling `core.SetFailed` is common amongst both `annotation` and `comment` reporters. Is it possible for us to define that logic into the parent abstract class here and inherit it in the subclasses?

mattisdada replied:
Done

mitchspano commented on src/reporter/reporter.types.ts:
```diff
@@ -0,0 +1,103 @@
+
+export interface Reporter {
+  write(): void;
+  translate(
```
nit: The verb `translate` can mean many things and could be a little confusing to a first time reader. Can we change this to something like `translateViolationToReport` or similar?

mitchspano commented on src/reporter/comments-reporter.ts:
```diff
@@ -0,0 +1,170 @@
+
+const HIDDEN_COMMENT_PREFIX = "<!--sfdx-scanner-->";
+
+export class CommentsReporter extends BaseReporter<GithubComment> {
```
Can we add jsdoc formatted comments on this class and all of its methods?

mattisdada replied:
Done

mitchspano commented on src/reporter/annoations-reporter.ts:
```diff
@@ -0,0 +1,96 @@
+export const ERROR = "Error";
+export const RIGHT = "RIGHT";
+
+export class AnnotationsReporter extends BaseReporter<GithubAnnotation> {
```
Can we add jsdoc formatted comments on this class and all of its methods?

mattisdada replied:
Done

mitchspano commented on src/index.ts:
```diff
@@ -123,39 +107,35 @@ function filterFindingsToDiffScope(
-      if (!isInChangedLines(violation, relevantLines) && !scannerFlags.target) {
+      const isNotChangedLines = !isInChangedLines(violation, relevantLines);
+      const isEmptyTarget = !inputs.target;
+      if (isNotChangedLines && isEmptyTarget) {
```
This resolves the "comment on all findings" issue mentioned above correct?

This is a bit of a nit so bare with me, but the booleans here require some mental energy to understand. Perhaps we can make this a little clearer to the reader:

```TS
const reportOnTarget = inputs.target && !pullRequest; // ignore target if provided unnecessarily on pull request 
const inScopeOfPr = pullRequest && isInChangedLines(violation, relevantLines);

if (!reportOnAllFiles && !inScopeOfPr) {
  continue;
}
```

mitchspano commented on src/reporter/reporter.types.ts:
```diff
@@ -0,0 +1,103 @@
+
+type GithubCommentSide = "RIGHT";
+
+export abstract class BaseReporter<T> implements Reporter {
```
Can we add jsdoc formatted comments on this class and all of its methods?

mattisdada replied:
Done

mattisdada commented:
I believe I've addressed most/all of the comments. I removed the changes around the scanning scope as fixes for that seemed to have been merged into main and looks okay to me

codefriar commented:
@mattisdada - I don't suppose i could convince you to also add the ability for the comments to be added as code review comments? That way we users can select between Check Runs, Code Review and PR Comments?

mattisdada commented:
> > @mattisdada - I don't suppose i could convince you to also add the ability for the comments to be added as code review comments? That way we users can select between Check Runs, Code Review and PR Comments?
> 
> I'm happy to explore it :) Although I probably won't extend this PR for it, maybe as a follow up

I very quickly put together an untested variant that includes commit commenting:
- Branch: https://github.com/mattisdada/sfdx-scan-pull-request/tree/feature/code-comment
- Changes: https://github.com/mitchspano/sfdx-scan-pull-request/commit/fb1ad78a9fd813f15ed362c2e59cca93422d1190

Can be used like:
```yaml
    - name: SFDX Scanner
      uses: mattisdada/sfdx-scan-pull-request@feature/code-comment
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        report-mode: commit-comment
```

The thing most likely to break is getting the commit sha, but feel free to give it a shot. I haven't tested it and it's bed time for me :) 

If it works well and @mitchspano is happy with it, I'll include it in a follow up PR after this one is merged.

mitchspano reviewed (approved):
This is looking great - thank you for your contributions!!!

matthew-gladman-oua commented:
> I am experiencing an issue when trying to test this version of the action out in a demo repository of mine:
> 
> ```yaml
> name: Static Analysis
> on:
>   pull_request:
>     types: [opened, reopened, edited, synchronize]
>   workflow_dispatch:
> jobs:
>   analyze:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@v3
>         with:
>           fetch-depth: 0 d
>       - name: Install SFDX CLI and Scanner
>         run: |
>           npm install sfdx-cli
>           npx sfdx plugins:install @salesforce/sfdx-scanner
>       - name: Run SFDX Scanner - Report findings as comments
>         uses: mattisdada/sfdx-scan-pull-request@ts-port
>         with:
>           report-mode: check-runs
>           severity-threshold: 2
>           strictly-enforced-rules: '[{ "engine": "pmd", "category": "Performance", "rule": "AvoidDebugStatements" }]'
>         env:
>           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
> ```
> 
> <img alt="image" width="1054" src="https://user-images.githubusercontent.com/18402464/231319523-9ea2d477-6aab-4594-aa9f-0ec3a0a63556.png">
> 
> Is it possible to add some improved error handling when calling out to the REST API?

Hmm, that looks like an error in the compiled code, that isn't a stack trace but just the compiled code :| Bit weird

Happy to add some additional error handling though

mattisdada commented:
I added in generic error handling, do you want to try it again and see what error is coming from GitHub?

mitchspano commented:
Slightly improved, but not 100% certain what the error is. It looks like it is failing to call `core.setFailed` and something is potentially also failing with the check runs API callout:

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/18402464/231600556-dc195692-54ac-4d23-ba11-9953c28ecade.png">

```
Creating Check Runs using GitHub REST API...
    at new Promise (<anonymous>)
Error: An error occurred while trying to write to GitHub
    at o (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:6065)
    at AnnotationsReporter.write (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:6760)
    at /home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:5718
    at Generator.next (<anonymous>)
    at fulfilled (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:2189)
```

mattisdada commented:
> Slightly improved, but not 100% certain what the error is. It looks like it is failing to call `core.setFailed` and something is potentially also failing with the check runs API callout:
> 
> <img alt="image" width="1027" src="https://user-images.githubusercontent.com/18402464/231600556-dc195692-54ac-4d23-ba11-9953c28ecade.png">
> 
> ```
> Performing static code analysis on all of the files in the difference...
> Running scan...
> npx sfdx scanner:run --target ".github/workflows/main.yml,.gitignore,ClassWithIssues.cls,MyClass.cls" --json
> Filtering the findings to just the lines which are part of the pull request...
> TypeError: Cannot read properties of undefined (reading 'setFailed')
>     at AnnotationsReporter.checkHasHaltingError (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:11477)
>     at AnnotationsReporter.<anonymous> (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:7[29](https://github.com/mitchspano/testActions/actions/runs/4683476428/jobs/8298593869#step:4:30)3)
>     at Generator.next (<anonymous>)
>     at /home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:6[31](https://github.com/mitchspano/testActions/actions/runs/4683476428/jobs/8298593869#step:4:32)8
> Creating Check Runs using GitHub REST API...
>     at new Promise (<anonymous>)
> Error: An error occurred while trying to write to GitHub
>     at o (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:6065)
>     at AnnotationsReporter.write (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:6760)
>     at /home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:5718
>     at Generator.next (<anonymous>)
>     at fulfilled (/home/runner/work/_actions/mattisdada/sfdx-scan-pull-request/ts-port/dist/index.js:1:2189)
> Error: A serious error has been identified
> ```

Thanks for the report, this was a weird import problem, should be resolved now.

Changes: https://github.com/mitchspano/sfdx-scan-pull-request/pull/40/commits/38e18720f02d0b84a5ab65175675d996b08a08f7

mitchspano commented:
It looks like the import issue is resolved, and most things seem to be working as expected now, but there are a couple minor things that stood out to me:

```yml
name: Static Analysis
on:
  pull_request:
    types: [opened, reopened, edited, synchronize]
  workflow_dispatch:
jobs:
  analyze:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0 d
      - name: Install SFDX CLI and Scanner
        run: |
          npm install sfdx-cli
          npx sfdx plugins:install @salesforce/sfdx-scanner
      - name: Run SFDX Scanner - Report findings
        uses: mattisdada/sfdx-scan-pull-request@ts-port
        with:
          report-mode: check-runs
          severity-threshold: 1
          strictly-enforced-rules: '[{ "engine": "pmd", "category": "Performance", "rule": "AvoidDebugStatements" }]'
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

```

<img width="1440" alt="Screenshot 2023-04-22 at 4 45 26 AM" src="https://user-images.githubusercontent.com/18402464/233776661-9ded5783-1dc5-45eb-9b08-58a8bfa39adc.png">

1. There are two findings within the action's run - `analyze` and `sfdx-scanner`. We expect `analyze` as that is within the action's yml, but where does the `sfdx-scanner` come from? Is that part of the `check-runs` feature?
1. There are two sequential failure messages:
  - `"One or more errors have been identified within the structure of the code that will need to be resolved before continuing.`
    - This one is what we expect, but it has a preceding double quote.
  -  `A serious error has been identified`
     - This one should not be printed to the log; we explicitly removed it as part of #29.

mattisdada commented:
> It looks like the import issue is resolved, and most things seem to be working as expected now, but there are a couple minor things that stood out to me:
> 
> ```yaml
> name: Static Analysis
> on:
>   pull_request:
>     types: [opened, reopened, edited, synchronize]
>   workflow_dispatch:
> jobs:
>   analyze:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@v3
>         with:
>           fetch-depth: 0 d
>       - name: Install SFDX CLI and Scanner
>         run: |
>           npm install sfdx-cli
>           npx sfdx plugins:install @salesforce/sfdx-scanner
>       - name: Run SFDX Scanner - Report findings
>         uses: mattisdada/sfdx-scan-pull-request@ts-port
>         with:
>           report-mode: check-runs
>           severity-threshold: 1
>           strictly-enforced-rules: '[{ "engine": "pmd", "category": "Performance", "rule": "AvoidDebugStatements" }]'
>         env:
>           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
> ```
> 
> <img alt="Screenshot 2023-04-22 at 4 45 26 AM" width="1440" src="https://user-images.githubusercontent.com/18402464/233776661-9ded5783-1dc5-45eb-9b08-58a8bfa39adc.png">
> 
> 1. There are two findings within the action's run - `analyze` and `sfdx-scanner`. We expect `analyze` as that is within the action's yml, but where does the `sfdx-scanner` come from? Is that part of the `check-runs` feature?
> 2. There are two sequential failure messages:
> 
> * `"One or more errors have been identified within the structure of the code that will need to be resolved before continuing.`
>   
>   * This one is what we expect, but it has a preceding double quote.
> * `A serious error has been identified`
>   
>   * This one should not be printed to the log; we explicitly removed it as part of [Update Error Message to be more Explicit #29](https://github.com/mitchspano/sfdx-scan-pull-request/issues/29).

Updated both of these in this commit: https://github.com/mitchspano/sfdx-scan-pull-request/pull/40/commits/fd440c9b82b0426e8daee6251c24f03a37e7de27

mattisdada commented:
> 1. expect `analyze` as that is within the action's yml, but where does the `sfdx-scanner` come from? Is that part of the `check-runs` feature?

Ah this is from check-runs, it's defined on this line:
- https://github.com/mitchspano/sfdx-scan-pull-request/blob/fd440c9b82b0426e8daee6251c24f03a37e7de27/src/reporter/annoations-reporter.ts#L62

mitchspano reviewed (approved):
Thank you so much for these changes!
